### PR TITLE
memtier, controllers: split out dynamic page demotion into a controller.

### PR DESCRIPTION
**Pull request by klihub**
This patch series is the first step in generalizing and improving dynamic page demotion by
introducing a new `page-migration` controller and splitting out the core of the functionality
from the `memtier` policy to this new controller. With this patch series applied any policy
can activate page demotion for container by simply attaching demotion policy decisions to
the desired containers.

In particular, the patch series
- adds functions for getting/setting page-migration policy/preferences to cache.Container
- adds a new `page-migration` controller
- moves the core of the demotion functionality to the new controller
- only leaves demotion/migration policy decision making in `memtier`
- fixes `resource-manager` to execute controller post-stop hooks for deleted containers

The current page-migration policy decision is implicitly 'idle page migration', with the policy
only providing the set of source and target nodes. Detecting idle pages and migrating them
from sources to targets is entirely taken care of in the page-migration controller. This should
be fairly easy to generalize in the future.

### CODE REVIEW

File: pkg/cri/resource-manager/cache/cache.go [hunk 1]
```diff
@@ -446,6 +455,26 @@ type Device struct {
 	Permissions string
 }
 
+// PageMigrate contains the policy/options for container page migration.
+type PageMigrate struct {
+	SourceNodes system.IDSet // idle memory pages on these NUMA nodes
+	TargetNodes system.IDSet // should be migrated to these NUMA nodes
+}
+
+func (pm *PageMigrate) Clone() *PageMigrate {
```
Comment (marquiz) [inline pkg/cri/resource-manager/cache/cache.go:464]: nit: add a comment/docstring to the exported methods

File: pkg/cri/resource-manager/cache/container.go [hunk 2]
```diff
@@ -811,6 +811,15 @@ func (c *container) GetToptierLimit() int64 {
 	return c.ToptierLimit
 }
 
+func (c *container) SetPageMigration(p *PageMigrate) {
```
Comment (marquiz) [inline pkg/cri/resource-manager/cache/container.go:814]: nit: comment/docstring

File: pkg/cri/resource-manager/control/page-migrate/flags.go [hunk 3]
```diff
@@ -0,0 +1,42 @@
...
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package pagemigrate
+
+import (
+	"github.com/intel/cri-resource-manager/pkg/config"
+)
+
+// Options captures our conigurable controller parameters.
```
Comment (marquiz) [inline pkg/cri/resource-manager/control/page-migrate/flags.go:21]: s/Options/options/

File: pkg/cri/resource-manager/control/page-migrate/page-migrate.go [hunk 4]
```diff
@@ -0,0 +1,99 @@
...
+	return nil
+}
+
+// PostStopHook is the controller's post-stop hook.
+func (m *migration) PostStopHook(cc cache.Container) error {
+	return nil
+}
+
+// migrationError creates a memory I/O-controller-specific formatted error message.
+func migrationError(format string, args ...interface{}) error {
```
Comment (marquiz) [inline pkg/cri/resource-manager/control/page-migrate/page-migrate.go:183]: memory I/O-controller??

File: pkg/cri/resource-manager/control/page-migrate/demoter.go [hunk 5]
```diff
@@ -217,7 +246,7 @@ func resetDirtyBit(pid string) error {
 }
 
 // ResetDirtyBit unsets soft-dirty bits for all processes in a container.
-func (d *demoter) ResetDirtyBit(c *container) error {
+func (d *demoter) resetDirtyBit(c *container) error {
```
Comment (marquiz) [inline pkg/cri/resource-manager/control/page-migrate/demoter.go:249]: nit: Change docstring correspondingly

File: pkg/cri/resource-manager/control/page-migrate/demoter.go [hunk 6]
```diff
@@ -97,32 +98,69 @@ func copyPagePool(p pagePool) pagePool {
...
+	if d.pageScanInterval > 0 && d.pageMoveInterval > 0 && d.maxPageMoveCount > 0 {
+		log.Info("scanning pages every %s, moving max. %d pages every %s",
+			d.pageScanInterval.String(), d.maxPageMoveCount, d.pageMoveInterval.String())
+		d.startDirtyBitResetTimer()
+	} else {
+		log.Info("scanning pages is disabled")
+	}
+}
+
+func (d *demoter) Stop() {
```
Comment (marquiz) [inline pkg/cri/resource-manager/control/page-migrate/demoter.go:119]: nit: I'd add comment to the exported methods. Or, alternatively, if demoter implements an interface state that in the demoter definition

**Review by marquiz — COMMENTED** [review]
Looks good, just a few very small nits.

One thing that puzzled me is why do we track containers locally in the controller? 🤔

**Comment by klihub** [comment]
> Looks good, just a few very small nits.
> 
> One thing that puzzled me is why do we track containers locally in the controller?

To avoid having to serialize access to the cache against request and event processing,
IOW against the resource-manager and policies. Or to put it another way to avoid having
to perform the potentially lengthy process of page scanning with the cache locked (which
then blocks request processing). 

An alternative would have been to duplicate what we had in the policy:
  - introduce controller events (basically callbacks invoked by the resource-manager with the cache locked)
  - periodically inject a controller event when we want to scan pages
  - scan pages from the event callback, with resource-manager holding the lock for us

Instead we now track a bare minimal copy of cached containers, just enough for doing
page scanning and migration completely on our own.

File: pkg/cri/resource-manager/cache/container.go (hunk 2 above)
Comment (klihub) [reply pkg/cri/resource-manager/cache/container.go:814]: It is present in the interface definition. Ditto for GetPageMigration().

File: pkg/cri/resource-manager/control/page-migrate/demoter.go (hunk 6 above)
Comment (klihub) [reply pkg/cri/resource-manager/control/page-migrate/demoter.go:120]: Added docstring to exported methods Stop() and Reconfigure().

File: pkg/cri/resource-manager/control/page-migrate/page-migrate.go (hunk 4 above)
Comment (klihub) [reply pkg/cri/resource-manager/control/page-migrate/page-migrate.go:183]: Fixed.

File: pkg/cri/resource-manager/control/page-migrate/flags.go (hunk 3 above)
Comment (klihub) [reply pkg/cri/resource-manager/control/page-migrate/flags.go]: Updated. And fixed the typo, too... ;-)

File: pkg/cri/resource-manager/cache/cache.go (hunk 1 above)
Comment (klihub) [reply pkg/cri/resource-manager/cache/cache.go:465]: Added docstring.

File: pkg/cri/resource-manager/cache/container.go (hunk 2 above)
Comment (marquiz) [reply pkg/cri/resource-manager/cache/container.go:814]: Yes, of course, sorry about this

**Comment by marquiz** [comment]
> To avoid having to serialize access to the cache against request and event processing,
> IOW against the resource-manager and policies. Or to put it another way to avoid having
> to perform the potentially lengthy process of page scanning with the cache locked (which
> then blocks request processing).
> 
> An alternative would have been to duplicate what we had in the policy:
> 
> * introduce controller events (basically callbacks invoked by the resource-manager with the cache locked)
> * periodically inject a controller event when we want to scan pages
> * scan pages from the event callback, with resource-manager holding the lock for us
> 
> Instead we now track a bare minimal copy of cached containers, just enough for doing
> page scanning and migration completely on our own.

Thanks for the explanation, makes sense 👍 Do we have this documented somewhere? If not, I think it would be geed to add e.g. a short comment in the code explaining the reason we're doing this. Might make it easier to read/maintain code in the future. WDYT?

**Comment by klihub** [comment]
> Thanks for the explanation, makes sense  Do we have this documented somewhere? If not, I think it would be geed to add e.g. a short comment in the code explaining the reason we're doing this. Might make it easier to read/maintain code in the future. WDYT?

Yes, that's very much true. I'll add a comment explaining this to the code.

**Comment by klihub** [comment]
> Yes, that's very much true. I'll add a comment explaining this to the code.

Added a comment about why we copy a subset of the containers to the controller. While at it, also updated the code to use a set of functions mimicking the corresponding cache functions when dealing with locally cached containers.

**Review by marquiz — APPROVED** [review]
Wow, that comment is pretty exhaustive 😊 👍 

LGTM

**Pull request merged**